### DISCUSSION TITLE
How to properly set up abiTypes to work with ethers v6?

### DISCUSSION THREAD

**Discussion by BravoNatalie** [wagmi-dev/abitype#211] [Q&A] [answered]
We have been using Typechain until now, but aim to change to ABItype. Although we really like the proposition of ABItypes, we have found it difficult to set up our contracts using only the documentation.

Our goal is to use ABItype with ethers, and that's what we have so far:

```ts
import {BaseContract as BaseContractEthers } from 'ethers'

type ReadFunctionNames<TAbi extends Abi> = ExtractAbiFunctionNames<TAbi, 'pure' | 'view'>

type FunctionArgsTypes<TAbi extends Abi> = AbiParametersToPrimitiveTypes<
  ExtractAbiFunction<TAbi, ReadFunctionNames<TAbi>>['inputs'],
  'inputs'
>[0]

type FunctionReturnTypes<TAbi extends Abi> = AbiParametersToPrimitiveTypes<
  ExtractAbiFunction<TAbi, ReadFunctionNames<TAbi>>['outputs'],
  'outputs'
>[0]

type ContractMethods<TAbi extends Abi> = {
  [name in ReadFunctionNames<TAbi>]: (...args: Array<FunctionArgsTypes<TAbi>>) => Promise<FunctionReturnTypes<TAbi>>
}

class BaseContract extends BaseContractEthers {}

export type BaseContractType<TAbi extends Abi> = BaseContract & ContractMethods<TAbi>

export const getContract = <TAbi extends Abi>(
  address: string,
  abi: InterfaceAbi,
  provider: JsonRpcProvider,
): BaseContractType<TAbi> => {
  const contract = new BaseContract(address, abi, provider) as BaseContractType<TAbi>
  return contract
}

//usage:

const contract = getContract<typeof ContractABI>(address, ContractABI, provider)

console.log(await contract.functionName())

``` 

But this is not giving us exactly what we need. The input and output types are not correct.
What is the proper way to integrate ABITypes with ethers? Could you give us an example?

**Comment by Raiden1411** [comment, marked as answer]
You can look at our examples in the repository that you can find [here](https://github.com/wagmi-dev/abitype/blob/main/playgrounds/functions/src/types.ts)

Here is also a example to wrap it on ethers that I made a while ago.
![image](https://github.com/wagmi-dev/abitype/assets/67233402/f8aeabd5-b28f-4027-a65d-b08b3d5763a3)

Another recomendation that I suggest is looking at the source code from wagmi pre migration to viem. You can find that [here](https://github.com/wagmi-dev/wagmi/blob/0.12.x/packages/core/src/actions/contracts/getContract.ts) and [here](https://github.com/wagmi-dev/wagmi/blob/0.12.x/packages/core/src/types/contracts.ts)

I hope this helps!

**Comment by jtourkos** [comment]
Hi @BravoNatalie,

I'm facing the same issue, and even though @Raiden1411's answer was really helpful, I still need a bit of help finding a working solution.
Everything's working fine at compile time but fails at runtime.

The problem is that TS expects the args to be an array, but for ethers to work, the array needs to be spread (which gives a compilation error...).

Would it be possible to post a minimal working setup based on the accepted answer to see what I am missing?

Thank you.

**Edit**:  I just noticed that the answer wasn't marked by you, so you might also have the same errors.

**Reply by Raiden1411** [reply]
> Would it be possible to post a minimal working setup based on the accepted answer to see what I am missing?

Yes, please do so.

**Comment by Raiden1411** [comment]
Hello @BravoNatalie and @jtourkos we created a draft PR with a example implementation of a wrapper for ethers v6 using abitype.

https://github.com/wevm/abitype/pull/214

Hopefully this helps your use cases.

**Reply by jtourkos** [reply]
Thank you so much for this. Works perfectly.

**Reply by DefiCake** [reply]
Just a headsup (and suggestion for improvement) for anyone who landed here, I 'll spare you the setup: if you are using ethers 's `Addressable` (e.g. passing a wallet object or a contract object instead of its address), this will not work for you.

<img width="610" alt="image" src="https://github.com/wevm/abitype/assets/46090742/a000c3f0-b4df-44e7-bae3-435a798195ce">

I 'd try to put it myself but I think this is a limitation of abitype itself, not the solution per se.
